### ISSUE TITLE
AppInjector violates dagger-android guidelines

### ISSUE THREAD

**Issue by Maragues** [googlesamples/android-architecture-components#99]
According to [Dagger Android](https://google.github.io/dagger/android.html#when-to-inject), we should invoke AndroidInjection.inject before Activity's super.onCreate

> It is crucial to call AndroidInjection.inject() before super.onCreate() in an Activity, since the call to super attaches Fragments from the previous activity instance during configuration change, which in turn injects the Fragments. In order for the Fragment injection to succeed, the Activity must already be injected.

[GitHubBrowser's AppInjector](https://github.com/googlesamples/android-architecture-components/blob/master/GithubBrowserSample/app/src/main/java/com/android/example/github/di/AppInjector.java) is a convenient way to avoid having to invoke *AndroidInjection.inject(this)*, but since it waits for *Application.ActivityLifecycleCallbacks.onActivityCreated* and this callback is invoked from *onCreate*, it's violating this strong recommendation by the dagger team.

Unfortunately there's nothing like *onActivityInstantiated*, so I'd say that the only option is to add *AndroidInjection.inject(this)* to our base class or extend from [DaggerActivty](https://google.github.io/dagger/api/latest/dagger/android/DaggerActivity.html)

**Comment by cgollner** [comment]
I've noticed this as well, and would be interested to know if the Dagger docs are correct, or if what they mention is not that important.

**Comment by svasilinets** [comment]
TLDR; this works in our sample because we use support library fragments. 

 As you correctly noticed: 
_Application.ActivityLifecycleCallbacks.onActivityCreated_ is called during _Activity.onCreate_ . However, we use support library fragments, so our activity's super class is _FragmentActivity_. 
Order of actions in its _onCreate_ method is:
`
super.onCreate(...); // calls  ActivityCallbacks.onCreate
restoreFragments(...);  // creates support lib Fragments
`

**Comment by maragues-kolibree** [comment]
I don't see those lines in *support-v4 26.0.0-beta2*, but I'll trust your deeper knowledge of Android internals.

In any case, since this is an example for the community and many people simply copy&paste, it might be useful to add a disclaimer.

Thanks for the insight, I'm closing this issue

**Comment by Maragues** [comment]
I don't see those lines in *support-v4 26.0.0-beta2*, but I'll trust your deeper knowledge of Android internals.

In any case, since this is an example for the community and many people simply copy&paste, it might be useful to add a disclaimer.

Thanks for the insight, I'm closing this issue

(sorry for double post, that was my work account)

**Comment by svasilinets** [comment]
Yeah, that was oversimplified version of _onCreate_ in _FragmentActivity_. The actual call is _mFragments.restoreAllState_
